nototools/cmap_data.py

```diff
@@ -0,0 +1,218 @@
+    root.text = sfx
+    for elem in root:
+      elem.tail = sfx
+      _prettify(elem, indent)
+    elem.tail = sfx[:-2]
+
+
+def _read_meta(meta_node):
+  """Return a MetaData object for the 'meta' element."""
+  date = meta_node.get('date')
+  program = meta_node.get('program')
+  args = []
+  args_node = meta_node.find('args')
+  if args_node is not None:
+    for arg_node in args_node:
+      key = arg_node.tag
+      val = arg_node.get('val').strip()
+      args.append((key, val))
+  return MetaData(date, program, args)
+
+
+def _read_table(table_node):
+  """Return a TableData object for the 'table' element."""
+  header = []
+  rows = []
+  for node in table_node:
+    if node.tag == 'th':
+      if header:
+        raise ValueError('cannot handle multiple headers')
+      elif rows:
+        raise ValueError('encountered header after rows')
+      else:
+        header = node.text.strip()
+    elif node.tag == 'tr':
+      rows.append(node.text.strip())
+  return create_table(header, rows)
+
+
+def _read_tree(root):
+  """Return a CmapData object for the 'cmapdata' element."""
+  meta = _read_meta(root.find('meta'))
+  table = _read_table(root.find('table'))
+  return CmapData(meta, table)
+
+
+def _build_meta(metadata):
+  """Create an xml 'meta' element for the MetaData object."""
+  meta = ET.Element('meta', date=metadata.date, program=metadata.program)
+  if metadata.args:
+    args = ET.Element('args')
+    for k, v in metadata.args:
+      arg = ET.Element(k, {'val': v})
+      args.append(arg)
+    meta.append(args)
+  return meta
+
+
+def _build_table(tabledata):
+  """Create an xml 'table' element for the TableData object."""
+  table = ET.Element('table', nrows=str(len(tabledata.rows)))
+  if tabledata.header:
+    header = ET.Element('th')
+    header.text = ','.join(tabledata.header)
+    table.append(header)
+  for rowdata in tabledata.rows:
+    row = ET.Element('tr')
+    row_items = [getattr(rowdata, h, '') for h in tabledata.header]
+    row.text = ','.join(row_items)
+    table.append(row)
+  return table
+
+
+def _build_tree(cmap_data, pretty=False):
+  """Create an xml 'cmapdata' element for the CmapData object."""
+  root = ET.Element('cmapdata')
+  def opt_append(elem):
+    if elem != None:
+      root.append(elem)
+  opt_append(_build_meta(cmap_data.meta))
+  opt_append(_build_table(cmap_data.table))
+  if pretty:
+    _prettify(root)
+  return ET.ElementTree(element=root)
+
+
+def read_cmap_data_file(filename):
+  return _read_tree(ET.parse(filename).getroot())
+
+
+def read_cmap_data(text):
+  return _read_tree(ET.fromstring(text))
+
+
+def write_cmap_data_file(cmap_data, filename, pretty=False):
+  _build_tree(cmap_data, pretty).write(
+      filename, encoding='utf-8', xml_declaration=True)
+
+
+def write_cmap_data(cmap_data, pretty=False):
+  return ET.tostring(_build_tree(cmap_data, pretty).getroot(), encoding='utf-8')
+
+
+def create_metadata(program, args=[], date=datetime.date.today()):
```

Comment:
Though probably OK in this case, using mutable/dynamic default values can cause weird problems since they're evaluated at module load time (style guide has details).